src/Chainweb/Pact/PactService.hs

```diff
@@ -720,20 +707,29 @@ execNewBlock
     -> Miner
     -> BlockCreationTime
     -> PactServiceM cas PayloadWithOutputs
-execNewBlock mpAccess parentHeader miner creationTime = withDiscardedBatch $ do
-    withCheckpointerRewind (Just (bHeight, pHash)) "execNewBlock" $ \pdbenv -> do
+execNewBlock mpAccess parentHeader miner creationTime =
+  do
+    updateMP
+    newTrans <- withDiscardedBatch $ withCheckpointerRewind (Just (bHeight, pHash)) "preBlock" $ \pdbenv -> do
```

Comment:
Rewrite to avoid `withDiscardedBatch` / `withCheckpointerRewind` twice as discussed in chat -- you can just `getCheckpointer`, run your stuff, then `_cpDiscard cp >> _cpRestore cp (Just (bheight, pHash))` should reset your state to play the next block